### PR TITLE
feat(analytics): add consent mode v2

### PR DESCRIPTION
**Pull request by matinzd**
## Description

Adding consent mode v2 to firebase analytics.

Reference: https://developers.google.com/tag-platform/security/guides/app-consent

## Related Issues

Fixes https://github.com/firebase/flutterfire/issues/12148

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by russellwheatley** [comment]
Thanks for the PR, @matinzd. As mentioned in the issue: https://github.com/firebase/flutterfire/issues/12148, we're discussing internally the shape of the API before rolling out 👍

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart [hunk 1]
```diff
@@ -125,7 +125,15 @@ class FirebaseAnalytics extends FirebasePluginPlatform {
...
-  /// Sets the applicable end user consent state. 'default' value for 'adStorageConsentGranted' & 'analyticsStorageConsentGranted' is 'true'
+  /// Sets the applicable end user consent state.
+  /// By default, no consent mode values are set. 
+  ///
+  /// Default consents can be set according to the platform: 
+  /// - [iOS][1]
+  /// - [Android][2]
+  ///
+  /// [1]: https://developers.google.com/tag-platform/security/guides/app-consent?platform=ios#default-consent
+  /// [2]: https://developers.google.com/tag-platform/security/guides/app-consent?platform=android#default-consent
```
Comment (matinzd) [inline packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:136]: According to the comment here, native APIs needs to be adjusted.  
@russellwheatley  I know that you are internally discussing this, so if you think I can continue this PR let me know and I can adjust it, if not, you can just close the PR.

**Comment by matinzd** [comment]
I cannot run the example app since there is something wrong with the pod install. 

```
Semantic Issue (Xcode): Use of undeclared identifier 'FIRAnalytics'
Semantic Issue (Xcode): Use of undeclared identifier 'FIRConsentTypeAnalyticsStorage'
...
```

Although pod install succeeds, `FirebaseAnalytics` pod folder is empty. 

Screenshot 2024-02-07 at 15 29 40

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart (hunk 1 above)
Comment (matinzd) [reply packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:136]: May I message you on discord? @russellwheatley

**Comment by matinzd** [comment]
Updated cocoapods and example app is now working. For some reason, the pods where not being copied to the corresponding folder. Example app works both on iOS and Android.

**Comment by matinzd** [comment]
Any updates? @russellwheatley

**Comment by russellwheatley** [comment]
@matinzd - not as such. Will let you know when it has been decided 👍

**Comment by shanelic** [comment]
> @matinzd - not as such. Will let you know when it has been decided 👍

Is there an estimated time for the decision to approve the PR?

**Comment by Sese-Schneider** [comment]
@matinzd thank you already for your effort implementing this!
Are you also aiming to provide the functionallity to FlutterFire **WEB** with this PR? https://firebase.google.com/docs/reference/js/analytics#setconsent_1697027

**Comment by matinzd** [comment]
@Sese-Schneider 
IMO, it's better to merge iOS and Android first and then have the web implemented in a separate PR.

**Comment by matinzd** [comment]
@Sese-Schneider 

IMO, it's better to merge iOS and Android first and then have the web implemented in a separate PR. For now, it seems like it's not priority at all as this PR is officially one month old and all of our campaigns paused due to this being a blocker for us.

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart [hunk 2]
```diff
@@ -125,14 +125,26 @@ class FirebaseAnalytics extends FirebasePluginPlatform {
...
   Future<void> setConsent({
     bool? adStorageConsentGranted,
     bool? analyticsStorageConsentGranted,
+    bool? adPersonalizationSignalsConsentGranted,
+    bool? adUserDataConsentGranted,
```
Comment (russellwheatley) [inline packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:141]: We also need to add 3 other consents for web platform here. See: https://firebase.google.com/docs/reference/js/analytics.consentsettings?_gl=1*kcmshc*_up*MQ..*_ga*MTk0NzA2OTc0Ni4xNzEwMjY0MDEx*_ga_CW55HF8NVT*MTcxMDI2NDAxMC4xLjAuMTcxMDI2NDAxMC4wLjAuMA..

File: packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart [hunk 3]
```diff
@@ -67,6 +67,8 @@ class FirebaseAnalyticsWeb extends FirebaseAnalyticsPlatform {
   Future<void> setConsent({
     bool? adStorageConsentGranted,
     bool? analyticsStorageConsentGranted,
+    bool? adPersonalizationSignalsConsentGranted,
+    bool? adUserDataConsentGranted,
```
Comment (russellwheatley) [inline packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart:71]: 3 additional consents mentioned earlier need to be updated here for web.

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Looking good, just need to pipe through the additional web consents and then I'll see what else needs to be updated, if anything 👍

**Comment by matinzd** [comment]
I am gonna adjust the PR tonight. Thanks for the feedback!

**Comment by russellwheatley** [comment]
@matinzd - we are looking to get this released next week, let me know if you have the time to complete, if not, I can take over. Thanks 😄

**Comment by matinzd** [comment]
Hey!
Will update it tonight.

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart (hunk 2 above)
Comment (matinzd) [reply packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:141]: Done

File: packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart [hunk 4]
```diff
@@ -67,6 +67,11 @@ class FirebaseAnalyticsWeb extends FirebaseAnalyticsPlatform {
...
+    bool? functionalityStorageConsentGranted,
+    bool? personalizationStorageConsentGranted,
+    bool? securityStorageConsentGranted,
   }) async {
     throw UnimplementedError('setConsent() is not supported on Web.');
```
Comment (russellwheatley) [inline packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart:76]: Could you set up `setConsent()` API for web, please? [Here is the documentation](https://firebase.google.com/docs/reference/js/analytics.md#setconsent_1697027)

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Ideally, we could get this done by some point tomorrow at the latest so we can include it in the latest release. Thanks

File: packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart (hunk 4 above)
Comment (matinzd) [reply packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart]: Hope this is ok. 

https://github.com/firebase/flutterfire/pull/12298/commits/45f02c60582b79f5d21dc499844ca12927f0143e

I need to test it now ...

**Comment by matinzd** [comment]
Fixed some formatting issues. Can you rerun the workflows @russellwheatley ?

**Comment by matinzd** [comment]
Rebased again. I reverted the change for firestore formatting.

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart [hunk 5]
```diff
@@ -125,14 +125,27 @@ class FirebaseAnalytics extends FirebasePluginPlatform {
...
```
Comment (russellwheatley) [inline packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:141]: You still haven't allowed all the arguments to be passed through to web. I would also make it clear in the inline documentation (i.e above the `setConsent()` function here and in the platform interface) which parameters are available for each platform.

**Review by russellwheatley — CHANGES_REQUESTED** [review]
@matinzd - Please pipe through the arguments to the web delegate and update inline documentation as mentioned. Thanks.

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart (hunk 5 above)
Comment (matinzd) [reply packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:151]: Done. 7abec92

**Comment by matinzd** [comment]
I guess it's ready now.

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart [hunk 6]
```diff
@@ -125,14 +125,44 @@ class FirebaseAnalytics extends FirebasePluginPlatform {
...
   }
 
-  /// Sets the applicable end user consent state. 'default' value for 'adStorageConsentGranted' & 'analyticsStorageConsentGranted' is 'true'
+  /// Sets the applicable end user consent state.
+  /// By default, no consent mode values are set.
+  ///
+  /// - [adStorageConsentGranted] - Enables storage, such as cookies, related to advertising. (Platform: Android, iOS, Web)
+  /// - [analyticsStorageConsentGranted] - Enables storage, such as cookies, related to analytics (for example, visit duration). (Platform: Android, iOS, Web)
+  /// - [adPersonalizationSignalsConsentGranted] - Sets consent for personalized advertising. (Platform: Android, iOS, Web)
+  /// - [adUserDataConsentGranted] - Sets consent for sending user data to Google for advertising purposes. (Platform: Android, iOS)
```
Comment (russellwheatley) [inline packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:134]:
```suggestion
  /// - [adUserDataConsentGranted] - Sets consent for sending user data to Google for advertising purposes. (Platform: Android, iOS, Web)
```

Comment (russellwheatley) [reply packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart:134]: Could you also add the documentation to above the same method in the platform interface: https://github.com/firebase/flutterfire/blob/master/packages/firebase_analytics/firebase_analytics_platform_interface/lib/src/platform_interface/platform_interface_firebase_analytics.dart#L170

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Nearly there! Thank you for the effort 👍

File: packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart (hunk 6 above)
Comment (matinzd) [reply packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart]: Done cf97b73.

Comment (matinzd) [reply packages/firebase_analytics/firebase_analytics/lib/src/firebase_analytics.dart]: f7774a3

**Comment by nitinmmnova** [comment]
@matinzd 

I've enabled 4 consents in the firebase analytics code but in the "Verbose" log we can see only 3 consents granted. I did not see "adPersonalization" consent property.

Do you know why?

How to enable verbose log you can check this [link](https://firebase.google.com/docs/analytics/events?platform=android#view_events_in_the_android_studio_debug_log)?

image
"

**Comment by nitinrgadhiya** [comment]
@matinzd 

FYI, I copied your code from PR.

I've enabled 4 consents in the Firebase analytics code but in the "Verbose" log I can see only 3 consents granted. I did not see "adPersonalization" consent property.

Do you know why?

How to enable a verbose log you can check this [link](https://firebase.google.com/docs/analytics/events?platform=android#view_events_in_the_android_studio_debug_log)?

image
"

**Comment by matinzd** [comment]
Which consent is missing in verbose log?

**Comment by nitinrgadhiya** [comment]
> adPersonalization

**Comment by nitinrgadhiya** [comment]
> > adPersonalization

@matinzd

**Comment by matinzd** [comment]
I couldn't find any issues with `adPersonalizationSignalsConsentGranted` key. It seems everything is in order. Keys are being properly passed into the platform method channel and on the native side, we have the same key as well. I am feeling that it's not showing up in the verbose log or it's just a user property which you can see in the log.

**Comment by nitinrgadhiya** [comment]
> I couldn't find any issues with `adPersonalizationSignalsConsentGranted` key. It seems everything is in order. Keys are being properly passed into the platform method channel and on the native side, we have the same key as well. I am feeling that it's not showing up in the verbose log or it's just a user property which you can see in the log.

yes, I agree with what you say ("about key passed properly."). And if your feeling is right, then It's okay. 
Thanks for quick response. @matinzd

**Comment by matinzd** [comment]
@russellwheatley ?

**Review by russellwheatley — APPROVED** [review]
LGTM

File: packages/firebase_analytics/firebase_analytics_web/lib/interop/analytics_interop.dart [hunk 7]
```diff
@@ -32,6 +32,12 @@ external void logEvent(
   JSObject? callOptions,
 );
 
+@JS()
+@staticInterop
+external void setConsent(
+  JSAny? consentSettings,
```
Comment (Lyokone) [inline packages/firebase_analytics/firebase_analytics_web/lib/interop/analytics_interop.dart:38]:
```suggestion
 // https://firebase.google.com/docs/reference/js/analytics.consentsettings.md#consentsettings_interface
  JSAny? consentSettings,
```

**Comment by russellwheatley** [comment]
Hey @matinzd - I just realised the integration tests are failing and need updating here: https://github.com/firebase/flutterfire/blob/master/tests/integration_test/firebase_analytics/firebase_analytics_e2e_test.dart#L211-L228

Update to this should work:

```dart
test(
      'setConsent',
      () async {
        await expectLater(
          FirebaseAnalytics.instance.setConsent(
            analyticsStorageConsentGranted: true,
            adStorageConsentGranted: true,
            adPersonalizationSignalsConsentGranted: true,
            adUserDataConsentGranted: true,
            functionalityStorageConsentGranted: true,
            personalizationStorageConsentGranted: true,
            securityStorageConsentGranted: true,
          ),
          completes,
        );
      },
    );
```

**Comment by matinzd** [comment]
Fixed. If anything else popped up, feel free to push to the branch 👍

**Review by Lyokone — APPROVED** [review]

**Review by russellwheatley — APPROVED** [review]

**Pull request merged**